nlp/model/augmented_lagrangian.py

```diff
@@ -128,6 +129,21 @@ def hess(self, *args, **kwargs):
         return self.hop(*args, **kwargs)
 
 
+    def magical_step(self, x, g):
+        """Compute a magical step for the slack variables.
+
+        This step minimizes the augmented Lagrangian with respect to the slack
+        variables only for a fixed set of decision variables. Here, g is the 
+        gradient computed by self.grad at x.
+        """
+        model = self.model
+        on = model.original_n
+        m_step = np.zeros(self.n)
+        m_step[on:] = -g[on:] / self.penalty
+        m_step = projected_step(x, m_step, model.Lvar, model.Uvar)
+        return m_step
+
+
```

Comment:
Actually, the problem is separable:

min  &Sigma;<sub>i</sub> (&frac12; s<sub>i</sub><sup>2</sup> + (c(x) - y/&rho;)<sub>i</sub> s<sub>i</sub>)
s.t. &ell;<sub>i</sub> ≤ s<sub>i</sub> ≤ u<sub>i</sub>

If you call i-th term in the sum q<sub>i</sub>(s<sub>i</sub>), the i-th component of the solution is s<sub>i</sub><sup>&star;</sup> := y<sub>i</sub>/&rho; - c<sub>i</sub>(x) if &ell;<sub>i</sub> ≤  s<sub>i</sub><sup>&star;</sup> ≤ u<sub>i</sub>, or else either &ell;<sub>i</sub> or u<sub>i</sub> (whichever produces the lowest value of q<sub>i</sub>(s<sub>i</sub>)).